### PR TITLE
Feature/message functionality

### PR DESCRIPTION
**Pull request by DeepankumarTpDev**
### **PR Description: Message Functionality and Conversation Handling**

This PR introduces the core message functionality, enabling users to communicate with ad posters via a messaging system. Key updates include:

- **Message Button on Ad Detail Page**: Users can send messages directly to ad posters.
- **Chat System**: A chat instance is created for each conversation between users for a specific ad and the Message Model stores each messages in the conversation.
- **Conversation Views**: Added views and templates for listing, displaying conversations and views for Editing and Deleting messages.
- **JavaScript Integration**: Client-side handling for message editing and deleting using Alpine.js.

This consolidates the messaging features for easier communication between users.

### CODE REVIEW

File: ads/views.py [hunk 1]
```diff
@@ -39,6 +41,30 @@ class AdDetailView(DetailView):
...
+
+        if self.ad.user.id > self.request.user.id:
+            user1, user2 = self.request.user, self.ad.user
+        else:
+            user1, user2 = self.ad.user, self.request.user
+
+        conversation_id = f"{user1.id}-{user1.username}-{self.ad.slug}-{user2.id}-{user2.username}"
+
+        user_conversations = get_user_conversations(request.user)
+        conversation_ids = [conv['conversation_id'] for conv in user_conversations]
```
Comment (shantanutomarr) [inline ads/views.py]: Not able to understand. Make it readable, extract it to another meaningful method. One should be able to understand what is happening here.

Comment (DeepankumarTpDev) [reply ads/views.py]: ok sir

File: chat/views.py [hunk 2]
```diff
@@ -1,3 +1,97 @@
...
+
+        return redirect('chat:conversation_detail', chat_id)
+    
+    def dispatch(self, request, *args, **kwargs):
+        if request.method == 'POST':
+            if 'edit' in request.path:
+                return self.edit_message(request, kwargs['chat_id'], kwargs['message_id'])
+            elif 'delete' in request.path:
+                print("delete")
+                return self.delete_message(request, kwargs['chat_id'], kwargs['message_id'])
```
Comment (shantanutomarr) [inline chat/views.py]: Please use separate views for each operation.

File: chat/views.py [hunk 3]
```diff
@@ -1,3 +1,97 @@
...
+        chat = get_object_or_404(Chat, id=chat_id, users=self.request.user)
+        message = get_object_or_404(Message, id=message_id, sender=request.user, chat=chat)
+
+        if request.method == "POST":
+            message.delete()
+
+        ismessagesinCurrentConversation = Message.objects.filter(chat=chat)
+
+        if not ismessagesinCurrentConversation:
+            return redirect('chat:conversation_list')
```
Comment (shantanutomarr) [inline chat/views.py]: Not able to understand

**Review by shantanutomarr — CHANGES_REQUESTED** [review]

**Review by shantanutomarr — APPROVED** [review]

**Pull request merged**